Mage/src/main/java/mage/abilities/effects/common/PutOnLibrarySourceEffect.java

```diff
@@ -47,12 +47,12 @@ public boolean apply(Game game, Ability source) {
         switch (game.getState().getZone(source.getSourceId())) {
             case BATTLEFIELD:
                 Permanent permanent = game.getPermanent(source.getSourceId());
-                if (permanent != null) {
+                if (permanent != null && source.getSourceObjectZoneChangeCounter() == permanent.getZoneChangeCounter(game)) {
                     result |= permanent.moveToZone(Zone.LIBRARY, source.getSourceId(), game, onTop);
                 }
             case GRAVEYARD:
                 Card card = game.getCard(source.getSourceId());
-                if (card != null) {
+                if (card != null && source.getSourceObjectZoneChangeCounter() == card.getZoneChangeCounter(game)) {
```

Comment:
Use standard call `source.getSourceObjectIfItStillExists(game)` instead manually ZCC compare.

Reply:
And there are same effects need to fix too with same code, search regexp: `class.+library.*source`
![shot_200209_214630](https://user-images.githubusercontent.com/8344157/74107066-c40ad800-4b85-11ea-9bc0-a0aa955de835.png)